Feat/ota 298/hmi api

patrickvacek opened the pull request:
There's a bit of cleaning up that could be done with the EventsInterpreter, and there is still some documentation to be done, but I'd like to solicit feedback on what I've got so far sooner rather than later.

lbonn commented on src/libaktualizr/primary/aktualizr.cc:
```diff
@@ -28,25 +23,30 @@ Aktualizr::Aktualizr(Config &config) : config_(config) {
+
+void Aktualizr::sendCommand(std::shared_ptr<command::BaseCommand> command) { *commands_channel_ << command; }
+
+void Aktualizr::setSignalHandler(std::function<void(std::shared_ptr<event::BaseEvent>)> &handler) {
```
It could be useful for the api user to remove a signal handler.

One possibility: return the `boost::signal2::connection` returned by `connect()`, so that `disconnect()` can be called from there.

cajun-rat replied:
How bad would it be to directly expose the boost::signals2::signal object?  We'd be binding our API to boost, but on the other hand we'd get all the logic like automatic disconnection for free.

patrickvacek replied:
It's an option. We are already bound to boost if I return the connection object (as I have now changed it to do). Sharing the signal object would expose `std::shared_ptr<boost::signals2::signal<void(std::shared_ptr<event::BaseEvent>)>>` instead of what I currently have (`boost::signals2::connection setSignalHandler(std::function<void(std::shared_ptr<event::BaseEvent>)>& handler);`), which is already a bit of a handful. I was considering ways to `typedef`/`using` that into something nicer, but I'm not sure what that would be that would actually make anyone's life easier. (Doing it for the function object is an option, but honestly it is probably better to know upfront what the required input and outputs of the handler are.)

OYTIS commented:
Looks good, so just download/installation progress is left and we can start with the actual HMI.

tkfu commented on src/libaktualizr/crypto/crypto.cc:
```diff
@@ -322,6 +322,7 @@ bool Crypto::extractSubjectCN(const std::string &cert, std::string *cn) {
 
 /**
  * Generate a RSA keypair
+ * @param key_type Type of key (RSA2048, RSA4096, etc.)
```
Can we actually enumerate the possible values here, instead of "etc."? I bet there are few enough that it's reasonable, right?

patrickvacek replied:
Yes, that is possible, although I would argue it would be better to go to the actual definitions directly instead of duplicating the information here. (Also, this was just my quick attempt to clean up some doxygen complaints -- this isn't part of the API that I wanted to work on.)

tkfu replied:
Ah, gotcha. And I think I agree that it's better not to.

patrickvacek commented:
Rebased three times, fixed the CI complaint, and fixed the doc comment. Should be ready now.

@OYTIS progress should be easy to implement (from the API perspective). It is likely that I will want to evolve the API next sprint. It should be usable and extensible now, but I think we can do even better.

lbonn commented on src/libaktualizr/primary/eventsinterpreter.cc:
```diff
@@ -18,29 +22,46 @@ EventsInterpreter::~EventsInterpreter() {
 void EventsInterpreter::interpret() { thread = std::thread(std::bind(&EventsInterpreter::run, this)); }
 
-std::shared_ptr<command::BaseCommand> EventsInterpreter::handle_cycle(event::BaseEvent &event, bool forever) {
+std::shared_ptr<command::BaseCommand> EventsInterpreter::handle_cycle(event::BaseEvent &event,
```
Is something preventing us from defining a new method? I kept `once` and `full` because they are really similar but `none` seems sufficiently different and simple:

```
std::shared_ptr<command::BaseCommand> EventsInterpreter::handle_none(event::BaseEvent &event) {
  if (event.variant == "SendDeviceDataComplete") {
    return std::make_shared<command::FetchMeta>();
  } else if (event.variant == "FetchMetaComplete") {
    return std::make_shared<command::CheckUpdates>();
  }
  return nullptr;
}
```

patrickvacek replied:
Done. Good call.
